### PR TITLE
Reactive

### PR DESCRIPTION
**Pull request by VolodymyrM27**
@OleksandRomaniuk

### CODE REVIEW

File: src/main/kotlin/systems/ajax/motrechko/airguardian/controller/DroneController.kt [hunk 1]
```diff
@@ -23,29 +26,25 @@ class DroneController(
     private val droneService: DroneService
 ) {
     @GetMapping
-    fun getAllDrone(): ResponseEntity<List<DroneResponse>> = ResponseEntity.ok(
-        droneService.getAllDrones().toResponse()
-    )
+    fun getAllDrone(): Flux<DroneResponse> =
+        droneService.getAllDrones().map { it.toResponse() }
+
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/controller/DroneController.kt:31]:
```suggestion
    fun getAllDrone(): Flux<DroneResponse> =
        droneService.getAllDrones().map { it.toResponse() }
```

File: src/main/kotlin/systems/ajax/motrechko/airguardian/controller/DroneController.kt [hunk 2]
```diff
@@ -23,29 +26,25 @@ class DroneController(
     private val droneService: DroneService
 ) {
     @GetMapping
-    fun getAllDrone(): ResponseEntity<List<DroneResponse>> = ResponseEntity.ok(
-        droneService.getAllDrones().toResponse()
-    )
+    fun getAllDrone(): Flux<DroneResponse> =
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/controller/DroneController.kt:29]: how does it look for a client? One response with a list of DroneResponse or many responses?
If many responses, then consider changing the return type to Mono<List<DroneResponse>> as the goal of this endpoint is to return all elements and complete

File: src/main/kotlin/systems/ajax/motrechko/airguardian/repository/DroneCustomReactiveRepository.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import systems.ajax.motrechko.airguardian.enums.DroneStatus
+import systems.ajax.motrechko.airguardian.model.Drone
+
+interface DroneCustomReactiveRepository {
+    fun findAllDronesWhereTheRemainingBatteryChargeIsLessThanAndHaveTheStatuses(
+        batteryLevel: Double,
+        statusesList: List<DroneStatus>
+    ): Flux<Drone>
+
+    fun findByStatus (droneStatus: DroneStatus): Flux<Drone>
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/repository/DroneCustomReactiveRepository.kt:15]:
```suggestion
    fun findAllByStatus (droneStatus: DroneStatus): Flux<Drone>
```
as method returns many items

File: src/main/kotlin/systems/ajax/motrechko/airguardian/service/DeliveryOrderService.kt [hunk 4]
```diff
@@ -3,62 +3,74 @@ package systems.ajax.motrechko.airguardian.service
...
-        } else {
-            order.status = DeliveryStatus.IN_PROGRESS
-            deliveryOrderRepository.save(order)
-            logger.info("no free drones were found for prayer with id {}", order.id)
-            order
-        }
+    fun createNewOrder(order: DeliveryOrder): Mono<DeliveryOrder> {
+        return droneLogisticsService.findAvailableDrones(order.items)
+            .collectList()
+            .filter { availableDrones -> availableDrones.isNotEmpty() }
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/service/DeliveryOrderService.kt:26]:
```suggestion
            .filter { it.isNotEmpty() }
```
don't afraid to use 'it' when naming is not definitely needed. here it's easy to understand that you're checking that list is not empty

File: src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneService.kt [hunk 5]
```diff
@@ -1,30 +1,35 @@
...
 
-    fun createDrone(drone: Drone): Drone = droneRepository.save(drone)
+    fun deleteDroneById(id: String): Mono<Unit> {
+        return droneRepository
+            .findById(id)
+            .switchIfEmpty(Mono.error(DroneNotFoundException("Drone with id $id not found")))
+            .flatMap {
+                droneRepository.deleteById(id)
+            }
+            .then(Mono.empty())
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneService.kt:29]: if your method is completed normally, don't return empty mono, instead return Mono.just(Unit)

**Review by ajax-shulepa-a — CHANGES_REQUESTED** [review]

File: src/main/kotlin/systems/ajax/motrechko/airguardian/controller/DroneController.kt (hunk 1 above)
Comment (VolodymyrM27) [reply src/main/kotlin/systems/ajax/motrechko/airguardian/controller/DroneController.kt:31]: In this suggestion no changes

Comment (ajax-shulepa-a) [reply src/main/kotlin/systems/ajax/motrechko/airguardian/controller/DroneController.kt:31]: remove extra empty line

File: src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneLogisticsService.kt [hunk 6]
```diff
@@ -83,21 +77,30 @@ class DroneLogisticsService(
...
         logger.debug("distance for delivery is {}", distance)
 
-        drone.status = DroneStatus.BUSY
-        drone.batteryLevel -= batteryConsumption
-        drone.flightHistory += FlightRecordUtils.createFlightRecord(
-            currentTime,
-            randomStartPosition,
-            order.deliveryCoordinates
-        )
+        drone.apply {
```
[the remark this answers is not in the document]
Comment (VolodymyrM27) [reply src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneLogisticsService.kt:101]: Is it mandatory, or is it a best practice for writing code?

[the remark this answers is not in the document]
Comment (ajax-shulepa-a) [reply src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneLogisticsService.kt:109]: for now, it is a best practice

File: src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneBatteryCheckService.kt [hunk 7]
```diff
@@ -22,18 +23,20 @@ class DroneBatteryCheckService(
...
-            drone.id, drone.batteryLevel
-        )
+    private fun createApplicationForCharging(drone: Drone): Mono<Unit> {
         drone.status = DroneStatus.NEED_TO_CHARGE
-        droneMongoRepository.save(drone)
+        return droneMongoRepository.save(drone)
+            .doOnSuccess{   logger.info(
+                "A request has been received to charge a drone with an ID {}, which is now has {}%",
+                drone.id, drone.batteryLevel
+            )}
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneBatteryCheckService.kt:37]:
```suggestion
            .doOnSuccess{  
             logger.info(
                "A request has been received to charge a drone with an ID {}, which is now has {}%",
                drone.id, drone.batteryLevel
            )
            }
```

File: src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneService.kt [hunk 8]
```diff
@@ -21,15 +25,20 @@ class DroneService(
 
     fun deleteDroneById(id: String): Mono<Unit> {
         return droneRepository
-            .findById(id)
-            .switchIfEmpty(Mono.error(DroneNotFoundException("Drone with id $id not found")))
-            .flatMap {
-                droneRepository.deleteById(id)
+            .deleteById(id)
+            .flatMap { deletedResult ->
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneService.kt:29]: replace with handle operator

File: src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneService.kt [hunk 9]
```diff
@@ -21,15 +25,20 @@ class DroneService(
...
+                }
             }
-            .then(Mono.empty())
     }
 
-    fun findDroneByStatus(droneStatus: DroneStatus): Flux<Drone> = droneRepository.findByStatus(droneStatus)
+    fun findDroneByStatus(droneStatus: DroneStatus): Mono<List<DroneResponse>> =
+        droneRepository.findAllByStatus(droneStatus)
+            .collectList()
+            .map { drones -> drones.map { it.toResponse() } }
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/systems/ajax/motrechko/airguardian/service/DroneService.kt:41]:
```suggestion
        droneRepository.findAllByStatus(droneStatus)
            .map { it.toResponse() }
            .collectList()
```
let's have mapping reactive

**Review by ajax-shulepa-a — APPROVED** [review]
Fix minor issues, please

**Pull request merged**